Add --include test

dangodev opened the pull request:
Adds test for the `--include` flag.

🚨Note that currently this test **does not pass** (see `node_modules` question below). 🚨

@FredKSchott I think I understand now what you were talking about with the `node_modules` directory. Yeah it feels a bit gross to commit those to version control. For each test suite, could we simply run `npm install` before `npm run TEST`? Those types of installs typically won’t slow down CI too much, and it prevents contributors from having to download all these `node_modules` for tests they may not need to run. So anyway, to get this to pass I think there are 2 paths forward:

1. I add the `node_modules` dir for this test to Git (and every subsequent test would have to, too)
2. We add an `npm install` command to `runner.js` before `npm run TEST` (saves VC at the expense of slightly-slower CI)

What do you think?

dangodev commented on __tests__/runner.js:
```diff
@@ -1,11 +1,14 @@
 
+function stripBenchmark(stdout) {
+  return stdout.replace(/\s*\[\d+s\]$/, ''); //remove benchmark
+}
```
We needed some way to remove the benchmarks, because fractions of seconds will throw the test. Is this OK?

FredKSchott replied:
I'd made a change so that the benchmark was never output when `NODE_ENV === 'test'`, but I like this method much better than my custom test-only behavior. Nice thinking. Can you replace my NODE_ENV workaround so that all tests match?

https://github.com/pikapkg/web/blob/4cba2484e84b62fb9ada17d3da5f684a1f87c076/src/index.ts#L419

dangodev commented on __tests__/runner.js:
```diff
@@ -39,10 +42,12 @@ for (const testName of readdirSync(__dirname)) {
-        readFileSync(path.join(entry.path1, entry.name1), {encoding: 'utf8'}),
-        readFileSync(path.join(entry.path2, entry.name2), {encoding: 'utf8'}),
-      );
+      if (entry.type1 === 'file') {
```
When @pika/web generates folders (`import httpVueLoader from '/web_modules/http-vue-loader/src/httpVueLoader.js'`), Node would throw an error trying to read a directory.

FredKSchott replied:
Ah, you mean that if the two folders were found to have not matched, we'd get a folder object here instead? That makes sense, although it would be nice to still throw an assertion error if two directories don't match. 

Also since you're here, can we turn the `noDiffSet` option on for these tests? I was just re-reading the https://www.npmjs.com/package/dir-compare#use docs and it looks like it will speed up our tests.

dangodev replied:
> can we turn the noDiffSet option on for these tests?

Oh nice! Didn’t know about that option. I think that would help.

But in general, I think we still need to have a check here because what’s breaking is `fs.readFileSync` trying to read a directory, so this should probably stay.

FredKSchott replied:
Right, I think I'm imagining something like this:

```
if (entry.type1 === 'file') {
  // read files, and assert.
} else if (entry.type1 === 'directory') {
  // don't try to read files, just assert.
}
```

Although that assertion would have less info than the file assertion error message. I guess, are we guarenteed that a folder diff mismatch always contains a file diff mismatch? In that case, then I'd just rewrite to here to be a bit more explicit to readers:


```
if (entry.type1 !== 'file') {
  // NOTE: We only compare files so that we give the test runner a more detailed diff.
  return;
} 
// read files, and assert.
```

dangodev replied:
Seems like the `noDiffSet` just disables the `diffSet` entirely (so we don’t get the nice output here as expected).

Enabling that would basically change the end of the test to:

```js
    var res = dircompare.compareSync(actualWebDependenciesLoc, expectedWebDependenciesLoc, {
      compareSize: true,
      noDiffSet: true,
    });
    return assert(res.same);
```

On my machine turning that off/on I couldn’t find any change in the time it took to run the tests (at least with the current number).

But it seems like the `diffSet`, no matter what, will always be an array of **every file it scanned** and not only the diffs (somewhat of a misnomer). On each item there’s a `state: 'equal'` that shows you 2 files/dirs are the same, and they’re in the `diffSet` only as proof they were scanned.

Anyway, I’m thinking we should probably leave everything as-is but just change the conditional to your last suggestion?

FredKSchott replied:
Ah dang, okay thanks for looking into it. +1 for keeping as is

FredKSchott commented:
Regarding `npm install` vs. checking node_modules into test: Since npm is this living, breathing ecosystem I'd prefer to have a test suite not rely on it. npm can go down, packages can change unexpectedly, harder to reproduce without knowing what code another developer installed, more complexity, etc.

I think a shared, checked in node_modules/ directory might be a good compromise. It would reduce the # of files checked in AND ensure that we aren't only testing @pika/web against simple 1-3 dependency node_modules graphs. That would also be easier to re-generate any time we want, as well.

If we want to add tests that specifically test "can ${PACKAGE_X} work with @pika/web" then I think that's a scenario where live-npm-install tests would make more sense. Since we do want to be notified ASAP if a new react versions breaks, for example.

dangodev commented:
Added the `node_modules` folder to VC, updated the test. Seeing they’re all passing locally (do we want to enable for pull requests)?

FredKSchott commented:
oh interesting, I'd assumed that they were running for PRs. Didn't realize node.yml and pull_request.yml were different? Hmm, strange. 

Are you familiar enough with Github Actions to make that change in this PR? Would love to get tests running everwhere now that we have them

dangodev commented:
> Are you familiar enough with Github Actions to make that change in this PR?

Yeah, actually, I’ve done a lot of CI setup. The filenames are arbitrary; you can name a workflow anything. The event (`on: []`) is what matters.

`on: [push]` means these tests will run on every code push on any branch, which is good but won’t block anything because they’ll only run after every push & merge. `on: [pull_request]` means that these checks will only be run on a PR, and can block merging if they don’t pass.

Anything like lint & tests that should be blocking should probably be on `pull_request` IMO. There are reasons to use every event, but I’ve found `pull_request` to be the most useful. It guarantees `master` is passing all checks.

All that to say, should we combine `lint` / `build` / and `test` jobs to all run in one file on `pull_request` to ensure every PR passes checks? That’s what I’d do.

dangodev commented:
<img width="775" alt="Screen Shot 2019-11-21 at 19 10 09" src="https://user-images.githubusercontent.com/1369770/69392221-8f783600-0c92-11ea-903a-23eeac7465fc.png">

This is what all the checks look like running on `pull_request`. Just for demonstration, I deleted the `nodejs.yml` workflow and moved everything into `pull_request.yml` (IMO “Node” isn’t a workflow, like branching, publishing, deploying, etc. would be). Now this will block PRs from breaking existing tests & functionality.

You can see that Node `8.x` failed; I think that’s because `require('fs').promises` doesn’t exist in `8.x`. So it’s working as expected?

FredKSchott commented:
Awesome! I actually removed v8 support on master earlier today, since it looks like we've never supported it due to our use of `fs.promises`. I'll remove it from your branch here so that we get passing on this PR.

dangodev commented:
Merge away! (I can’t merge myself). This looks good. Yay, tests! 🎉

FredKSchott commented:
Fixed! You should now be able to push.

Thanks for the awesome work, as always!
